Src/Fido2/AttestationFormat/Apple.cs

```diff
@@ -0,0 +1,148 @@
+                if (DateTime.Now < x5ccert.NotBefore || DateTime.Now > x5ccert.NotAfter)
+                    throw new Fido2VerificationException("Apple signing certificate expired or not yet valid");
+            }
+
+            // The attestation certificate attestnCert MUST be the first element in the array.
+            var attestnCert = new X509Certificate2(X5c.Values.First().GetByteString());
+
+            // Version MUST be set to 3
+            if (3 != attestnCert.Version)
+                throw new Fido2VerificationException("Apple x5c attestation certificate not V3");
+
+            // The Basic Constraints extension MUST have the CA component set to false
+            if (IsAttnCertCACert(attestnCert.Extensions))
+                throw new Fido2VerificationException("Attestation certificate has CA cert flag present");
+
+            // Verify that the x5c array contains the intermediate and leaf certificates for atttestation,
+            // starting from the credential certificate stored in the first data buffer in the array.
+            // Verify the validity of the certificates using Apple's Webauthn Root CA
+            var trustPath = X5c.Values
+                .Select(x => new X509Certificate2(x.GetByteString()))
+                .ToArray();
+
+            // If the authenticator is listed as in the metadata as one that should produce a basic full attestation, build and verify the chain
+            var entry = _metadataService?.GetEntry(AuthData.AttestedCredentialData.AaGuid);
+
+            // while conformance testing, we must reject any authenticator that we cannot get metadata for
+            if (_metadataService?.ConformanceTesting() == true && null == entry)
+                throw new Fido2VerificationException("AAGUID not found in MDS test metadata");
+
+            // If the authenticator is listed as in the metadata as one that should produce a basic full attestation, build and verify the chain
+            if (entry?.MetadataStatement?.AttestationTypes.Contains((ushort)MetadataAttestationType.ATTESTATION_BASIC_FULL) ?? false)
+            {
+                var attestationRootCertificates = entry.MetadataStatement.AttestationRootCertificates
+                    .Select(x => new X509Certificate2(Convert.FromBase64String(x)))
+                    .ToArray();
+
+                if (false == ValidateTrustChain(trustPath, attestationRootCertificates))
+                {
+                    throw new Fido2VerificationException("Invalid certificate chain in apple attestation");
+                }
+            }
+
+            // 2. Create clientDataHash as the SHA256 hash of the client data, and append that hash to the end of the authenticator
+            // data (authData from the decoded object)
+            // 3. Generate SHA-256 hash of the concatenation of authData and clientDataHash, i.e. SHA256(authData || SHA256(clientData))
+            var expectedNonce = SHA256.Create().ComputeHash(Data);

```

Comment:
Is `Data` already the concatenation mentioned in the comments? If not, the comments doesn't seem to match the code.

Reply:
Data is the concatenation mentioned in the comments.